libraries/AP_NavEKF/EKFGSF_yaw.cpp

```diff
@@ -0,0 +1,678 @@
+
+ 	// inv(S) * innovation
+	float tempVec[2];
+	tempVec[0] = invMat[0][0] * EKF[mdl_idx].innov[0] + invMat[0][1] * EKF[mdl_idx].innov[1];
+	tempVec[1] = invMat[1][0] * EKF[mdl_idx].innov[0] + invMat[1][1] * EKF[mdl_idx].innov[1];
+
+	// transpose(innovation) * inv(S) * innovation
+	float normDist = tempVec[0] * EKF[mdl_idx].innov[0] + tempVec[1] * EKF[mdl_idx].innov[1];
+
+    // convert from a normalised variance to a probability assuming a Gaussian distribution
+	normDist = expf(-0.5f * normDist);
+	normDist *= sqrtf(t4)/ M_2PI;
+	return normDist;
+}
+
+bool EKFGSF_yaw::getLogData(float *yaw_composite, float *yaw_composite_variance, float yaw[N_MODELS_EKFGSF], float innov_VN[N_MODELS_EKFGSF], float innov_VE[N_MODELS_EKFGSF], float weight[N_MODELS_EKFGSF])
+{
+	if (vel_fuse_running) {
+		memcpy(yaw_composite, &GSF.state[2], sizeof(GSF.state[2]));
+		memcpy(yaw_composite_variance, &GSF.yaw_variance, sizeof(GSF.yaw_variance));
+		for (uint8_t mdl_idx = 0; mdl_idx < N_MODELS_EKFGSF; mdl_idx++) {
+			yaw[mdl_idx] = EKF[mdl_idx].X[2];
+			innov_VN[mdl_idx] = EKF[mdl_idx].innov[0];
+			innov_VE[mdl_idx] = EKF[mdl_idx].innov[1];
+			weight[mdl_idx] = GSF.weights[mdl_idx];
+		}
+		return true;
+	}
+	return false;
+}
+
+void EKFGSF_yaw::forceSymmetry(const uint8_t mdl_idx)
+{
+	float P01 = 0.5f * (EKF[mdl_idx].P[0][1] + EKF[mdl_idx].P[1][0]);
+	float P02 = 0.5f * (EKF[mdl_idx].P[0][2] + EKF[mdl_idx].P[2][0]);
+	float P12 = 0.5f * (EKF[mdl_idx].P[1][2] + EKF[mdl_idx].P[2][1]);
+	EKF[mdl_idx].P[0][1] = EKF[mdl_idx].P[1][0] = P01;
+	EKF[mdl_idx].P[0][2] = EKF[mdl_idx].P[2][0] = P02;
+	EKF[mdl_idx].P[1][2] = EKF[mdl_idx].P[2][1] = P12;
+}
+
+// Apply a body frame delta angle to the body to earth frame rotation matrix using a small angle approximation
+Matrix3f EKFGSF_yaw::updateRotMat(const Matrix3f &R, const Vector3f &g)
+{
+	Matrix3f ret = R;
+	ret[0][0] += R[0][1] * g[2] - R[0][2] * g[1];
+	ret[0][1] += R[0][2] * g[0] - R[0][0] * g[2];
+	ret[0][2] += R[0][0] * g[1] - R[0][1] * g[0];
+	ret[1][0] += R[1][1] * g[2] - R[1][2] * g[1];
+	ret[1][1] += R[1][2] * g[0] - R[1][0] * g[2];
+	ret[1][2] += R[1][0] * g[1] - R[1][1] * g[0],
+    ret[2][0] += R[2][1] * g[2] - R[2][2] * g[1];
+	ret[2][1] += R[2][2] * g[0] - R[2][0] * g[2];
+	ret[2][2] += R[2][0] * g[1] - R[2][1] * g[0];
+
+	// Renormalise rows
+	float rowLengthSq;
+	for (uint8_t r = 0; r < 3; r++) {
+		rowLengthSq = ret[r][0] * ret[r][0] + ret[r][1] * ret[r][1] + ret[r][2] * ret[r][2];
+		if (rowLengthSq > FLT_EPSILON) {
+			// Use linear approximation for inverse sqrt taking advantage of the row length being close to 1.0
+			const float rowLengthInv = 1.5f - 0.5f * rowLengthSq;
+			ret[r][0] *= rowLengthInv;
+			ret[r][1] *= rowLengthInv;
+			ret[r][2] *= rowLengthInv;
+		}
+    }
+
+	return ret;
+}
+
+bool EKFGSF_yaw::getYawData(float *yaw, float *yawVariance)
+{
+	if (!vel_fuse_running) {
+		return false;
+	}
+	memcpy(yaw, &GSF.state[2], sizeof(GSF.state[2]));
```

Comment:
why memcpy for a float assignment?

Reply:
That was originally returning the GSF covariance matrix. Will fix

Reply:
Fixed